File: buildSrc/src/main/java/com/google/firebase/gradle/plugins/FirebaseJavaLibraryPlugin.java
// Copyright 2019 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.firebase.gradle.plugins;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableMap;
import com.google.firebase.gradle.plugins.apiinfo.ApiInformationTask;
import com.google.firebase.gradle.plugins.apiinfo.GenerateApiTxtFileTask;
import com.google.firebase.gradle.plugins.apiinfo.GetMetalavaJarTask;
import com.google.firebase.gradle.plugins.ci.Coverage;
import java.io.File;
import java.nio.file.Paths;
import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.attributes.Attribute;
import org.gradle.api.file.FileCollection;
import org.gradle.api.plugins.JavaPluginConvention;
import org.gradle.api.tasks.SourceSet;
import org.gradle.api.tasks.TaskProvider;
import org.jetbrains.kotlin.gradle.tasks.KotlinCompile;

// TODO(vkryachko): extract functionality common across Firebase{,Java}LibraryPlugin plugins.
public class FirebaseJavaLibraryPlugin implements Plugin<Project> {

  @Override
  public void apply(Project project) {
    project.apply(ImmutableMap.of("plugin", "java-library"));
    FirebaseLibraryExtension firebaseLibrary =
        project
            .getExtensions()
            .create("firebaseLibrary", FirebaseLibraryExtension.class, project, LibraryType.JAVA);

    // reduce the likelihood of kotlin module files colliding.
    project
        .getTasks()
        .withType(
            KotlinCompile.class,
            kotlin ->
                kotlin
                    .getKotlinOptions()
                    .setFreeCompilerArgs(
                        ImmutableList.of("-module-name", kotlinModuleName(project))));

    setupStaticAnalysis(project, firebaseLibrary);
    project.afterEvaluate(p -> Dokka.configure(project, null, firebaseLibrary));
  }

  private static void setupStaticAnalysis(Project project, FirebaseLibraryExtension library) {
    project.afterEvaluate(
        p ->
            project
                .getConfigurations()
                .all(
                    c -> {
                      if ("annotationProcessor".equals(c.getName())) {
                        for (String checkProject : library.staticAnalysis.errorproneCheckProjects) {
                          project
                              .getDependencies()
                              .add("annotationProcessor", project.project(checkProject));
                        }
                      }
                      if ("lintChecks".equals(c.getName())) {
                        for (String checkProject :
                            library.staticAnalysis.androidLintCheckProjects) {
                          project
                              .getDependencies()
                              .add("lintChecks", project.project(checkProject));
                        }
                      }
                    }));

    setupApiInformationAnalysis(project);

    project.getTasks().register("firebaseLint", task -> task.dependsOn("lint"));
    Coverage.apply(library);
  }

  private static void setupApiInformationAnalysis(Project project) {
    File metalavaOutputJarFile = new File(project.getRootProject().getBuildDir(), "metalava.jar");
    SourceSet mainSourceSet =
        project
            .getConvention()
            .getPlugin(JavaPluginConvention.class)
            .getSourceSets()
            .getByName("main");
    File outputFile =
        project
            .getRootProject()
            .file(
                Paths.get(
                    project.getRootProject().getBuildDir().getPath(),
                    "apiinfo",
                    project.getPath().substring(1).replace(":", "_")));
    File outputApiFile = new File(outputFile.getAbsolutePath() + "_api.txt");

    project
        .getTasks()
        .register(
            "getMetalavaJar",
            GetMetalavaJarTask.class,
            task -> {
              task.setOutputFile(metalavaOutputJarFile);
            });
    File apiTxt =
        project.file("api.txt").exists()
            ? project.file("api.txt")
            : project.file(project.getRootDir() + "/empty-api.txt");
    TaskProvider<ApiInformationTask> apiInfo =
        project
            .getTasks()
            .register(
                "apiInformation",
                ApiInformationTask.class,
                task -> {
                  task.setApiTxt(apiTxt);
                  task.setMetalavaJarPath(metalavaOutputJarFile.getAbsolutePath());
                  task.setSourceSet(mainSourceSet);
                  task.setOutputFile(outputFile);
                  task.setBaselineFile(project.file("baseline.txt"));
                  task.setOutputApiFile(outputApiFile);
                  if (project.hasProperty("updateBaseline")) {
                    task.setUpdateBaseline(true);
                  } else {
                    task.setUpdateBaseline(false);
                  }
                  task.dependsOn("getMetalavaJar");
                });

    TaskProvider<GenerateApiTxtFileTask> generateApiTxt =
        project
            .getTasks()
            .register(
                "generateApiTxtFile",
                GenerateApiTxtFileTask.class,
                task -> {
                  task.setApiTxt(project.file("api.txt"));
                  task.setMetalavaJarPath(metalavaOutputJarFile.getAbsolutePath());
                  task.setSourceSet(mainSourceSet);
                  task.setBaselineFile(project.file("baseline.txt"));
                  task.setUpdateBaseline(project.hasProperty("updateBaseline"));
                  task.dependsOn("getMetalavaJar");
                });

    TaskProvider<GenerateStubsTask> docStubs =
        project
            .getTasks()
            .register(
                "docStubs",
                GenerateStubsTask.class,
                task -> {
                  task.dependsOn("getMetalavaJar");

                  task.setSourceSet(mainSourceSet);
                });
    project.getTasks().getByName("check").dependsOn(docStubs);

    project.afterEvaluate(
        p -> {
          FileCollection classpath =
              project
                  .getConfigurations()
                  .getByName("runtimeClasspath")
                  .getIncoming()
                  .artifactView(
                      config ->
                          config.attributes(
                              container ->
                                  container.attribute(
                                      Attribute.of("artifactType", String.class), "jar")))
                  .getArtifacts()
                  .getArtifactFiles();

          apiInfo.configure(t -> t.setClassPath(classpath));
          generateApiTxt.configure(t -> t.setClassPath(classpath));
          docStubs.configure(t -> t.setClassPath(classpath));
        });
  }

  private static String kotlinModuleName(Project project) {

    String fullyQualifiedProjectPath = project.getPath().replaceAll(":", "-");

    return project.getRootProject().getName() + fullyQualifiedProjectPath;
  }
}

File: buildSrc/src/main/java/com/google/firebase/gradle/plugins/FirebaseJavaLibraryPlugin.kt
/*
 * Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.firebase.gradle.plugins

import com.google.common.collect.ImmutableList
import com.google.firebase.gradle.plugins.LibraryType.JAVA
import com.google.firebase.gradle.plugins.semver.ApiDiffer
import com.google.firebase.gradle.plugins.semver.GmavenCopier
import org.gradle.api.Project
import org.gradle.api.attributes.Attribute
import org.gradle.api.plugins.JavaLibraryPlugin
import org.gradle.api.plugins.JavaPluginConvention
import org.gradle.kotlin.dsl.apply
import org.gradle.kotlin.dsl.create
import org.gradle.kotlin.dsl.getPlugin
import org.gradle.kotlin.dsl.register
import org.gradle.kotlin.dsl.withType
import org.jetbrains.kotlin.gradle.tasks.KotlinCompile

class FirebaseJavaLibraryPlugin : BaseFirebaseLibraryPlugin() {

  override fun apply(project: Project) {
    project.apply<JavaLibraryPlugin>()
    project.apply<DackkaPlugin>()

    setupFirebaseLibraryExtension(project)
    registerMakeReleaseNotesTask(project)

    // reduce the likelihood of kotlin module files colliding.
    project.tasks.withType<KotlinCompile> {
      kotlinOptions.freeCompilerArgs = ImmutableList.of("-module-name", kotlinModuleName(project))
    }
  }

  private fun setupFirebaseLibraryExtension(project: Project) {
    val firebaseLibrary =
      project.extensions.create<FirebaseLibraryExtension>("firebaseLibrary", project, JAVA)

    setupStaticAnalysis(project, firebaseLibrary)
    setupApiInformationAnalysis(project)
    getIsPomValidTask(project, firebaseLibrary)
    setupVersionCheckTasks(project, firebaseLibrary)
    configurePublishing(project, firebaseLibrary)
  }

  private fun setupVersionCheckTasks(project: Project, firebaseLibrary: FirebaseLibraryExtension) {
    project.tasks.register<GmavenVersionChecker>("gmavenVersionCheck") {
      groupId.value(firebaseLibrary.groupId.get())
      artifactId.value(firebaseLibrary.artifactId.get())
      version.value(firebaseLibrary.version)
      latestReleasedVersion.value(firebaseLibrary.latestReleasedVersion.orElseGet { "" })
    }
    project.mkdir("semver")
    project.tasks.register<GmavenCopier>("copyPreviousArtifacts") {
      dependsOn("jar")
      project.file("semver/previous.jar").delete()
      groupId.value(firebaseLibrary.groupId.get())
      artifactId.value(firebaseLibrary.artifactId.get())
      aarAndroidFile.value(false)
      filePath.value(project.file("semver/previous.jar").absolutePath)
    }
    val currentJarFile =
      project
        .file("build/libs/${firebaseLibrary.artifactId.get()}-${firebaseLibrary.version}.jar")
        .absolutePath
    val previousJarFile = project.file("semver/previous.jar").absolutePath
    project.tasks.register<ApiDiffer>("semverCheck") {
      currentJar.value(currentJarFile)
      previousJar.value(previousJarFile)
      version.value(firebaseLibrary.version)
      previousVersionString.value(
        GmavenHelper(firebaseLibrary.groupId.get(), firebaseLibrary.artifactId.get())
          .getLatestReleasedVersion()
      )

      dependsOn("copyPreviousArtifacts")
    }
  }

  private fun setupApiInformationAnalysis(project: Project) {
    val srcDirs =
      project.convention.getPlugin<JavaPluginConvention>().sourceSets.getByName("main").java.srcDirs

    val apiInfo = getApiInfo(project, srcDirs)

    val generateApiTxt = getGenerateApiTxt(project, srcDirs)
    val docStubs = getDocStubs(project, srcDirs)

    project.tasks.getByName("check").dependsOn(docStubs)
    project.afterEvaluate {
      val classpath =
        configurations
          .getByName("runtimeClasspath")
          .incoming
          .artifactView {
            attributes { attribute(Attribute.of("artifactType", String::class.java), "jar") }
          }
          .artifacts
          .artifactFiles
      apiInfo.configure { classPath = classpath }
      generateApiTxt.configure { classPath = classpath }
      docStubs.configure { classPath = classpath }
    }
  }
}
